File: _backup_/20260726-1635/src/base/reportes.ts
// YvexPOS Móvil — Lógica de reportes de ventas (locales).
//
// Todos los rangos y agrupaciones se calculan en HORA LOCAL del teléfono
// (la venta de las 11pm cuenta en el día local correcto, no en el día UTC).
// creado_en se guarda en ISO UTC; los límites de rango se calculan en JS y la
// agrupación por día/hora también, para no depender del parseo de fechas de
// SQLite.

import { bd } from "./db";

export type Rango = "hoy" | "7d" | "30d" | "mes";

/** ISO UTC del inicio (00:00 local) de hace `dias` días. */
function inicioDiaLocalISO(diasAtras: number): string {
  const d = new Date();
  d.setDate(d.getDate() - diasAtras);
  d.setHours(0, 0, 0, 0);
  return d.toISOString();
}

/** ISO UTC del inicio del rango elegido. */
export function inicioDeRango(rango: Rango): string {
  if (rango === "hoy") return inicioDiaLocalISO(0);
  if (rango === "7d") return inicioDiaLocalISO(6);
  if (rango === "30d") return inicioDiaLocalISO(29);
  // mes: día 1 del mes actual
  const d = new Date();
  d.setDate(1);
  d.setHours(0, 0, 0, 0);
  return d.toISOString();
}

// ---------------------------------------------------------------------------
// Resumen del rango
// ---------------------------------------------------------------------------

export type ResumenRango = {
  total_centavos: number;
  tickets: number;
  promedio_centavos: number;
  articulos: number;
};

export async function resumenRango(rango: Rango): Promise<ResumenRango> {
  const db = await bd();
  const desde = inicioDeRango(rango);
  const fila = await db.getFirstAsync<{ n: number; t: number }>(
    `SELECT COUNT(*) AS n, COALESCE(SUM(total_centavos),0) AS t
     FROM ventas WHERE estado <> 'cancelada' AND creado_en >= ?`,
    [desde]
  );
  const art = await db.getFirstAsync<{ a: number }>(
    `SELECT COALESCE(SUM(l.cantidad),0) AS a
     FROM venta_lineas l JOIN ventas v ON v.id = l.venta_id
     WHERE v.estado <> 'cancelada' AND v.creado_en >= ?`,
    [desde]
  );
  const n = fila?.n ?? 0;
  const t = fila?.t ?? 0;
  return {
    total_centavos: t,
    tickets: n,
    promedio_centavos: n > 0 ? Math.round(t / n) : 0,
    articulos: art?.a ?? 0,
  };
}

// ---------------------------------------------------------------------------
// Serie para la gráfica: por HORA (rango hoy) o por DÍA (los demás)
// ---------------------------------------------------------------------------

export type PuntoSerie = {
  etiqueta: string; // "14h" o "9 jul"
  total_centavos: number;
  tickets: number;
};

export async function serieVentas(rango: Rango): Promise<PuntoSerie[]> {
  const db = await bd();
  const desde = inicioDeRango(rango);
  const filas = await db.getAllAsync<{ creado_en: string; total_centavos: number }>(
    `SELECT creado_en, total_centavos FROM ventas
     WHERE estado <> 'cancelada' AND creado_en >= ?`,
    [desde]
  );

  if (rango === "hoy") {
    // Cubetas por hora local, de 0 a la hora actual.
    const horaActual = new Date().getHours();
    const cubetas: PuntoSerie[] = [];
    for (let h = 0; h <= horaActual; h++) {
      cubetas.push({ etiqueta: `${h}h`, total_centavos: 0, tickets: 0 });
    }
    for (const f of filas) {
      const h = new Date(f.creado_en).getHours();
      if (h >= 0 && h <= horaActual) {
        cubetas[h].total_centavos += f.total_centavos;
        cubetas[h].tickets += 1;
      }
    }
    return cubetas;
  }

  // Cubetas por día local desde el inicio del rango hasta hoy.
  const inicio = new Date(desde);
  const hoy = new Date();
  hoy.setHours(0, 0, 0, 0);
  const cubetas: PuntoSerie[] = [];
  const indice = new Map<string, number>(); // clave "yyyy-m-d" -> posición
  const cursor = new Date(inicio);
  while (cursor <= hoy) {
    const clave = `${cursor.getFullYear()}-${cursor.getMonth()}-${cursor.getDate()}`;
    indice.set(clave, cubetas.length);
    cubetas.push({
      etiqueta: cursor.toLocaleDateString("es-MX", { day: "numeric", month: "short" }),
      total_centavos: 0,
      tickets: 0,
    });
    cursor.setDate(cursor.getDate() + 1);
  }
  for (const f of filas) {
    const d = new Date(f.creado_en);
    const clave = `${d.getFullYear()}-${d.getMonth()}-${d.getDate()}`;
    const i = indice.get(clave);
    if (i !== undefined) {
      cubetas[i].total_centavos += f.total_centavos;
      cubetas[i].tickets += 1;
    }
  }
  return cubetas;
}

// ---------------------------------------------------------------------------
// Top productos del rango
// ---------------------------------------------------------------------------

export type TopProducto = {
  nombre: string;
  cantidad: number;
  total_centavos: number;
};

export async function topProductos(rango: Rango, limite = 8): Promise<TopProducto[]> {
  const db = await bd();
  return db.getAllAsync<TopProducto>(
    `SELECT l.nombre_producto AS nombre,
            COALESCE(SUM(l.cantidad),0) AS cantidad,
            COALESCE(SUM(l.total_linea_centavos),0) AS total_centavos
     FROM venta_lineas l
     JOIN ventas v ON v.id = l.venta_id
     WHERE v.estado <> 'cancelada' AND v.creado_en >= ?
     GROUP BY l.nombre_producto
     ORDER BY total_centavos DESC
     LIMIT ?`,
    [inicioDeRango(rango), limite]
  );
}

// ---------------------------------------------------------------------------
// Ventas por departamento
// ---------------------------------------------------------------------------

export type VentaDepto = {
  departamento: string;
  color: string | null;
  total_centavos: number;
  cantidad: number;
};

export async function ventasPorDepartamento(rango: Rango): Promise<VentaDepto[]> {
  const db = await bd();
  return db.getAllAsync<VentaDepto>(
    `SELECT COALESCE(c.nombre, 'Sin departamento') AS departamento,
            c.color AS color,
            COALESCE(SUM(l.total_linea_centavos),0) AS total_centavos,
            COALESCE(SUM(l.cantidad),0) AS cantidad
     FROM venta_lineas l
     JOIN ventas v ON v.id = l.venta_id
     LEFT JOIN productos p ON p.id = l.producto_id
     LEFT JOIN categorias c ON c.id = p.categoria_id
     WHERE v.estado <> 'cancelada' AND v.creado_en >= ?
     GROUP BY c.id
     ORDER BY total_centavos DESC`,
    [inicioDeRango(rango)]
  );
}

// ---------------------------------------------------------------------------
// Ventas por método de pago
// ---------------------------------------------------------------------------

export type VentaMetodo = {
  metodo: string;
  total_centavos: number;
  pagos: number;
};

export async function ventasPorMetodo(rango: Rango): Promise<VentaMetodo[]> {
  const db = await bd();
  return db.getAllAsync<VentaMetodo>(
    `SELECT p.metodo,
            COALESCE(SUM(p.monto_centavos),0) AS total_centavos,
            COUNT(*) AS pagos
     FROM pagos p
     JOIN ventas v ON v.id = p.venta_id
     WHERE v.estado <> 'cancelada' AND v.creado_en >= ?
     GROUP BY p.metodo
     ORDER BY total_centavos DESC`,
    [inicioDeRango(rango)]
  );
}

File: pc/yvexiq-pos/src-tauri/src/db/reportes.rs
//! Reportes (solo lectura). Métricas agregadas para el dueño/gerente.
//!
//! Toda la ganancia/margen usa el costo HISTÓRICO guardado en venta_lineas
//! (costo_unitario_centavos) para ser fiel a la realidad. Ventas previas a la
//! migración 004 tienen costo 0 y se reportan como "sin costo registrado".
//!
//! Filtra siempre por ventas no canceladas. El rango de fechas se recibe como
//! ISO (inicio y fin), calculado en el frontend según el periodo elegido.
//!
//! NOTA (sincronización): los reportes son del NEGOCIO COMPLETO. Las ventas
//! que bajan de otras cajas (móvil, otra PC) ya viven en la base local y
//! cuentan igual que las propias; excluirlas daría números a medias.
//! (El corte de caja es la excepción: ese sí es por turno, porque el
//! efectivo es físico de cada caja.)

use rusqlite::Connection;
use serde::Serialize;

#[derive(Debug, Serialize)]
pub struct MetricasGenerales {
    pub total_vendido_centavos: i64,
    pub num_ventas: i64,
    pub venta_promedio_centavos: i64,
    pub ganancia_centavos: i64,
    pub margen_promedio_pct: f64,
    pub costo_total_centavos: i64,
    pub articulos_vendidos: f64,
}

#[derive(Debug, Serialize)]
pub struct ParMetodo {
    pub metodo: String,
    pub monto_centavos: i64,
    pub num: i64,
}

#[derive(Debug, Serialize)]
pub struct FilaCategoria {
    pub categoria: String,
    pub vendido_centavos: i64,
    pub ganancia_centavos: i64,
    pub articulos: f64,
}

#[derive(Debug, Serialize)]
pub struct FilaProducto {
    pub nombre: String,
    pub cantidad: f64,
    pub vendido_centavos: i64,
    pub ganancia_centavos: i64,
}

#[derive(Debug, Serialize)]
pub struct PuntoTiempo {
    pub etiqueta: String, // fecha YYYY-MM-DD o hora 00-23
    pub vendido_centavos: i64,
    pub num_ventas: i64,
}

#[derive(Debug, Serialize)]
pub struct ReporteCompleto {
    pub metricas: MetricasGenerales,
    pub por_metodo: Vec<ParMetodo>,
    pub por_categoria: Vec<FilaCategoria>,
    pub productos_top: Vec<FilaProducto>,
    pub por_dia: Vec<PuntoTiempo>,
    pub por_hora: Vec<PuntoTiempo>,
}

/// Genera el reporte completo para un rango [inicio, fin] (ISO-8601 UTC).
/// Cuenta las ventas de TODAS las cajas del negocio (ver nota de arriba).
pub fn generar(
    con: &Connection,
    inicio: &str,
    fin: &str,
) -> Result<ReporteCompleto, String> {
    let metricas = metricas_generales(con, inicio, fin)?;
    let por_metodo = por_metodo(con, inicio, fin)?;
    let por_categoria = por_categoria(con, inicio, fin)?;
    let productos_top = productos_top(con, inicio, fin)?;
    let por_dia = por_dia(con, inicio, fin)?;
    let por_hora = por_hora(con, inicio, fin)?;
    Ok(ReporteCompleto {
        metricas,
        por_metodo,
        por_categoria,
        productos_top,
        por_dia,
        por_hora,
    })
}

fn metricas_generales(
    con: &Connection,
    inicio: &str,
    fin: &str,
) -> Result<MetricasGenerales, String> {
    // Total e ingresos.
    let (total, num): (i64, i64) = con
        .query_row(
            "SELECT COALESCE(SUM(total_centavos),0), COUNT(*)
             FROM ventas
             WHERE estado != 'cancelada'
               AND creado_en >= ?1 AND creado_en <= ?2",
            rusqlite::params![inicio, fin],
            |r| Ok((r.get(0)?, r.get(1)?)),
        )
        .map_err(|e| format!("error métricas: {e}"))?;

    // Ganancia y costo desde las líneas (costo histórico).
    // OJO: cantidad es REAL (granel), así que costo*cantidad sale decimal.
    // Leemos como f64 y redondeamos a centavos enteros.
    let (ingreso_lineas, costo_total_f, articulos): (i64, f64, f64) = con
        .query_row(
            "SELECT
               COALESCE(SUM(vl.total_linea_centavos),0),
               COALESCE(SUM(vl.costo_unitario_centavos * vl.cantidad),0),
               COALESCE(SUM(vl.cantidad),0)
             FROM venta_lineas vl
             JOIN ventas v ON vl.venta_id = v.id
             WHERE v.estado != 'cancelada'
               AND v.creado_en >= ?1 AND v.creado_en <= ?2",
            rusqlite::params![inicio, fin],
            |r| Ok((r.get(0)?, r.get(1)?, r.get(2)?)),
        )
        .map_err(|e| format!("error ganancia: {e}"))?;
    let costo_total = costo_total_f.round() as i64;

    let ganancia = ingreso_lineas - costo_total;
    let venta_promedio = if num > 0 { total / num } else { 0 };
    // Margen sobre ingreso de líneas (no sobre total con descuento global, para
    // que sea margen de producto). Si no hay ingreso, 0.
    let margen = if ingreso_lineas > 0 {
        (ganancia as f64 / ingreso_lineas as f64) * 100.0
    } else {
        0.0
    };

    Ok(MetricasGenerales {
        total_vendido_centavos: total,
        num_ventas: num,
        venta_promedio_centavos: venta_promedio,
        ganancia_centavos: ganancia,
        margen_promedio_pct: (margen * 10.0).round() / 10.0,
        costo_total_centavos: costo_total,
        articulos_vendidos: articulos,
    })
}

fn por_metodo(con: &Connection, inicio: &str, fin: &str) -> Result<Vec<ParMetodo>, String> {
    let mut stmt = con
        .prepare(
            "SELECT p.metodo, COALESCE(SUM(p.monto_centavos),0), COUNT(DISTINCT p.venta_id)
             FROM pagos p JOIN ventas v ON p.venta_id = v.id
             WHERE v.estado != 'cancelada'
               AND v.creado_en >= ?1 AND v.creado_en <= ?2
             GROUP BY p.metodo
             ORDER BY 2 DESC",
        )
        .map_err(|e| format!("error por_metodo: {e}"))?;
    let filas = stmt
        .query_map(rusqlite::params![inicio, fin], |r| {
            Ok(ParMetodo { metodo: r.get(0)?, monto_centavos: r.get(1)?, num: r.get(2)? })
        })
        .map_err(|e| format!("error por_metodo q: {e}"))?;
    filas.collect::<Result<_, _>>().map_err(|e| format!("error por_metodo c: {e}"))
}

fn por_categoria(con: &Connection, inicio: &str, fin: &str) -> Result<Vec<FilaCategoria>, String> {
    let mut stmt = con
        .prepare(
            "SELECT COALESCE(c.nombre, 'Sin categoría'),
                    COALESCE(SUM(vl.total_linea_centavos),0),
                    COALESCE(SUM(vl.total_linea_centavos - vl.costo_unitario_centavos * vl.cantidad),0),
                    COALESCE(SUM(vl.cantidad),0)
             FROM venta_lineas vl
             JOIN ventas v ON vl.venta_id = v.id
             LEFT JOIN productos p ON vl.producto_id = p.id
             LEFT JOIN categorias c ON p.categoria_id = c.id
             WHERE v.estado != 'cancelada'
               AND v.creado_en >= ?1 AND v.creado_en <= ?2
             GROUP BY COALESCE(c.nombre, 'Sin categoría')
             ORDER BY 2 DESC",
        )
        .map_err(|e| format!("error por_categoria: {e}"))?;
    let filas = stmt
        .query_map(rusqlite::params![inicio, fin], |r| {
            // ganancia sale decimal (cantidad REAL); leer f64 y redondear.
            let ganancia_f: f64 = r.get(2)?;
            Ok(FilaCategoria {
                categoria: r.get(0)?,
                vendido_centavos: r.get(1)?,
                ganancia_centavos: ganancia_f.round() as i64,
                articulos: r.get(3)?,
            })
        })
        .map_err(|e| format!("error por_categoria q: {e}"))?;
    filas.collect::<Result<_, _>>().map_err(|e| format!("error por_categoria c: {e}"))
}

fn productos_top(con: &Connection, inicio: &str, fin: &str) -> Result<Vec<FilaProducto>, String> {
    let mut stmt = con
        .prepare(
            "SELECT vl.descripcion,
                    COALESCE(SUM(vl.cantidad),0),
                    COALESCE(SUM(vl.total_linea_centavos),0),
                    COALESCE(SUM(vl.total_linea_centavos - vl.costo_unitario_centavos * vl.cantidad),0)
             FROM venta_lineas vl
             JOIN ventas v ON vl.venta_id = v.id
             WHERE v.estado != 'cancelada'
               AND v.creado_en >= ?1 AND v.creado_en <= ?2
             GROUP BY vl.descripcion
             ORDER BY 3 DESC
             LIMIT 15",
        )
        .map_err(|e| format!("error productos_top: {e}"))?;
    let filas = stmt
        .query_map(rusqlite::params![inicio, fin], |r| {
            let ganancia_f: f64 = r.get(3)?;
            Ok(FilaProducto {
                nombre: r.get(0)?,
                cantidad: r.get(1)?,
                vendido_centavos: r.get(2)?,
                ganancia_centavos: ganancia_f.round() as i64,
            })
        })
        .map_err(|e| format!("error productos_top q: {e}"))?;
    filas.collect::<Result<_, _>>().map_err(|e| format!("error productos_top c: {e}"))
}

fn por_dia(con: &Connection, inicio: &str, fin: &str) -> Result<Vec<PuntoTiempo>, String> {
    let mut stmt = con
        .prepare(
            "SELECT date(creado_en), COALESCE(SUM(total_centavos),0), COUNT(*)
             FROM ventas
             WHERE estado != 'cancelada'
               AND creado_en >= ?1 AND creado_en <= ?2
             GROUP BY date(creado_en)
             ORDER BY date(creado_en)",
        )
        .map_err(|e| format!("error por_dia: {e}"))?;
    let filas = stmt
        .query_map(rusqlite::params![inicio, fin], |r| {
            Ok(PuntoTiempo { etiqueta: r.get(0)?, vendido_centavos: r.get(1)?, num_ventas: r.get(2)? })
        })
        .map_err(|e| format!("error por_dia q: {e}"))?;
    filas.collect::<Result<_, _>>().map_err(|e| format!("error por_dia c: {e}"))
}

fn por_hora(con: &Connection, inicio: &str, fin: &str) -> Result<Vec<PuntoTiempo>, String> {
    // strftime %H da la hora (00-23) en UTC. Para Mazatlán habría que ajustar
    // zona; por ahora se reporta en la hora local del timestamp guardado.
    let mut stmt = con
        .prepare(
            "SELECT strftime('%H', creado_en), COALESCE(SUM(total_centavos),0), COUNT(*)
             FROM ventas
             WHERE estado != 'cancelada'
               AND creado_en >= ?1 AND creado_en <= ?2
             GROUP BY strftime('%H', creado_en)
             ORDER BY 1",
        )
        .map_err(|e| format!("error por_hora: {e}"))?;
    let filas = stmt
        .query_map(rusqlite::params![inicio, fin], |r| {
            Ok(PuntoTiempo { etiqueta: r.get(0)?, vendido_centavos: r.get(1)?, num_ventas: r.get(2)? })
        })
        .map_err(|e| format!("error por_hora q: {e}"))?;
    filas.collect::<Result<_, _>>().map_err(|e| format!("error por_hora c: {e}"))
}
